[agents docs] add pipelines.md etc (#13567)
huggingface/diffusers: .ai/pipelines.md

> # Pipeline conventions and rules

Author: added a pipeline doc for agent

   ```python
   # don't do this — @property for static config value
   @property
   def is_turbo(self) -> bool:
       return bool(getattr(self.transformer.config, "is_turbo", False))

   # don't do this — module-level constant duplicating loadable config
   SAMPLE_RATE = 48000

   # do this — set once in __init__ with a getattr fallback (see pipeline_flux.py:209)
   def __init__(self, ..., vae, transformer, ...):
       ...
       self.register_modules(vae=vae, transformer=transformer, ...)
       self.vae_scale_factor = (
           2 ** (len(self.vae.config.block_out_channels) - 1) if getattr(self, "vae", None) else 8
       )
       self.sample_rate = int(self.vae.config.sampling_rate) if getattr(self, "vae", None) else 48000
   ```
   `@property` is reserved for per-call state — values that depend on something set inside `__call__` (e.g. `do_classifier_free_guidance` reading `self._guidance_scale`).

2. **`@torch.no_grad()` discipline.** Two failure modes:
    - **Missing on `__call__` entirely** — causes GPU OOM from gradient accumulation during inference. Always decorate `__call__` with `@torch.no_grad()`.

>     - **Redundant inside helpers** that `__call__` already covers. The decorator puts every descendant in no-grad, so an inner `with torch.no_grad():` is noise — and worse, it forecloses callers who want to invoke `pipe.encode_prompt(...)` with grads enabled (training, embedding optimization). Convention across diffusers (flux, qwen, flux2, stable_audio, audioldm2) is decorator-only.

Reviewer: ```suggestion
    - **Redundant inside helpers** that `__call__` already covers. The decorator puts every descendent in no-grad, so an inner `with torch.no_grad():` is noise — and worse, it forecloses callers who want to invoke `pipe.encode_prompt(...)` with grads enabled (training, embedding optimization). Convention across diffusers (flux, qwen, flux2, stable_audio, audioldm2) is decorator-only.
```

> 3. **Reinventing logic that already exists in the repo.** Before adding new logic to a pipeline, check whether diffusers already has it. Reuse out of the box if you can; extend with a small kwarg if the change is genuinely small (use judgement on whether to extend an existing class or write a new one).

Reviewer: ```suggestion
3. **Reinventing logic that already exists in the repo.** Check `src/diffusers/guiders/` and `src/diffusers/schedulers/` before adding new logic. Reuse what's already there; extend with a small kwarg for minor variations.
```
